triton-constraint-circuit/src/lib.rs

```diff
@@ -1471,4 +1696,328 @@ mod tests {
+                assert!(
+                    builder.all_nodes.borrow().is_empty(),
+                    "fresh hashmap should be empty"
+                );
+
+                let mut num_main_inputs = 0;
+                let mut num_aux_inputs = 0;
+                let mut all_nodes = vec![];
+                let mut output_nodes = vec![];
+
+                for input in inputs {
+                    match input {
+                        CircuitInputType::Main => {
+                            let node = builder.input(II::main_table_input(num_main_inputs));
+                            num_main_inputs += 1;
+                            all_nodes.push(node);
+                        }
+                        CircuitInputType::Aux => {
+                            let node = builder.input(II::aux_table_input(num_aux_inputs));
+                            num_aux_inputs += 1;
+                            all_nodes.push(node);
+                        }
+                    }
+                }
+
+                for i in 0..num_challenges {
+                    let node = builder.challenge(i);
+                    all_nodes.push(node);
+                }
+
+                for constant in constants {
+                    match constant {
+                        CircuitConstantType::Base(bfe) => {
+                            let node = builder.b_constant(bfe);
+                            all_nodes.push(node);
+                        }
+                        CircuitConstantType::Extension(xfe) => {
+                            let node = builder.x_constant(xfe);
+                            all_nodes.push(node);
+                        }
+                    }
+                }
+
+                if !all_nodes.is_empty() {
+                    for operation in operations {
+                        match operation {
+                            CircuitOperationChoice::Add(lhs, rhs) => {
+                                let lhs_index = lhs % all_nodes.len();
+                                let rhs_index = rhs % all_nodes.len();
+
+                                let lhs_node = all_nodes[lhs_index].clone();
+                                let rhs_node = all_nodes[rhs_index].clone();
+
+                                let node = lhs_node + rhs_node;
+                                all_nodes.push(node);
+                            }
+                            CircuitOperationChoice::Mul(lhs, rhs) => {
+                                let lhs_index = lhs % all_nodes.len();
+                                let rhs_index = rhs % all_nodes.len();
+
+                                let lhs_node = all_nodes[lhs_index].clone();
+                                let rhs_node = all_nodes[rhs_index].clone();
+
+                                let node = lhs_node * rhs_node;
+                                all_nodes.push(node);
+                            }
+                        }
```

Comment:
Common code can be factored out.

```suggestion
                        let (lhs, rhs) = match operation {
                            CircuitOperationChoice::Add(lhs, rhs) => (lhs, rhs),
                            CircuitOperationChoice::Mul(lhs, rhs) => (lhs, rhs),
                        };
    
                        let lhs_index = lhs % all_nodes.len();
                        let rhs_index = rhs % all_nodes.len();
    
                        let lhs_node = all_nodes[lhs_index].clone();
                        let rhs_node = all_nodes[rhs_index].clone();
    
                        let node = match operation {
                            CircuitOperationChoice::Add(_, _) => lhs_node + rhs_node,
                            CircuitOperationChoice::Mul(_, _) => lhs_node * rhs_node,
                        };
                        all_nodes.push(node);
```